## Dataset inspection

In this part,we want to firstly do the dataset inspection, we import important package and check for the values in the dataset.

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
df = pd.read_csv('D:/wx/WeChat Files/GRT0848/FileStorage/File/2022-07/amazon_co-ecommerce_sample(1).csv')

Always good to check for repeating and null values in the dataset. As you can see our data is like a purchasing record and it includes many variables like "manufacturer", "numbers_in_stock", "number_of_review". We will firstly do the data inspection and then do the normalization, finally the ETL part. 

In [3]:
df=df.drop(['uniq_id','sellers','number_of_answered_questions','description'],axis=1)

We decided to further check the information of this dataset using this method. 

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 13 columns):
 #   Column                                       Non-Null Count  Dtype 
---  ------                                       --------------  ----- 
 0   product_name                                 10000 non-null  object
 1   manufacturer                                 9993 non-null   object
 2   price                                        8565 non-null   object
 3   number_available_in_stock                    7500 non-null   object
 4   number_of_reviews                            9982 non-null   object
 5   average_review_rating                        9982 non-null   object
 6   amazon_category_and_sub_category             9310 non-null   object
 7   customers_who_bought_this_item_also_bought   8938 non-null   object
 8   product_information                          9942 non-null   object
 9   product_description                          9349 non-null   object
 10  items_custo

As we can see, there exists 10000 rows in the dataframe. All columns have been assgined "object" datatype except for number_of_answered_questions that was interpreted as "numeric". Also, there exists some rows that don't have the whole data, like row "price" only have 8565 non-null rows, customer questions and answers only have 914 non-null rows. We will need to take those in to account when designing the database. We need to make sure which rows are accepetable to have Null values and which rows are not acceptable.

Then, we want to check for repeating values. Are there 10000 individual product names?  

In [5]:
df.product_name[df.product_name.duplicated()]

471                                       Polyhedral Dice
585              100 six sided dice, 14mm, random colours
686                                Barbie Advent Calendar
729               Schleich Farm Christmas Advent Calendar
730     Playmobil 5494 Christmas Advent Calendar Santa...
783              Schleich Horse Christmas Advent Calendar
1663    Set Of 12 Foam Water Pistol HYDRO STORM Blaste...
1669    INTEX Inflatable Swimming Paddling Play Pool 3...
1747    Konfidence Youth Swim Jacket - Blue/Navy - 10-...
3258    TAVERN WENCH ADULT FANCY DRESS COSTUME LADIES ...
4647    Zoo Animal Hand Sock Glove Finger Puppets Sack...
5036    Happy 6th Birthday Giant Party Wall Banner 3 B...
5082    Happy 18th Birthday Party Giant Wall Banner 3 ...
5765                        Power Trains Auto Loader City
5815    Meng "Model 1:35 Toyota Hilux Pick Up Truck w/...
6417    Optimus Prime DMK 01 Transformers Movie Dual M...
6551     Iron Man 3 Marvel Iron Assemblers Battle Vehicle
6647          

We can see from the result that there are indeed titles which are repeating like " Barbie Advent Calendar", "Schleich Horse Christmas Advent Calendar","Melissa & Doug Stacking Construction Vehicles". 

In [6]:
df.drop_duplicates(subset=None, keep='first', inplace=True)

## Normalization

In [7]:
df.insert(0, 'product_id', range(1, 1 + len(df)))

In [8]:
df['price'] = df['price'].str.extractall('(\d+)').unstack().fillna('').sum(axis=1).astype(int)
df['price'] = df['price'] * 0.01

In [9]:
dfcate = pd.DataFrame(df, columns=['amazon_category_and_sub_category'])
dfcate = dfcate.amazon_category_and_sub_category.str.split(pat='>',expand=True)
dfcate.head()

0                              1                2        3     4
0  Hobbies    Model Trains & Railway Sets    Rail Vehicles    Trains  None
1  Hobbies    Model Trains & Railway Sets    Rail Vehicles    Trains  None
2  Hobbies    Model Trains & Railway Sets    Rail Vehicles    Trains  None
3  Hobbies    Model Trains & Railway Sets    Rail Vehicles    Trains  None
4  Hobbies    Model Trains & Railway Sets    Rail Vehicles    Trains  None

In [10]:
dfcateg = pd.DataFrame(dfcate, columns=[0])
dfcateg.drop_duplicates(subset=None, keep='first', inplace=True)
dfcateg.insert(0, 'category_id', range(1, 1 + len(dfcateg)))
dfcateg = dfcateg.rename(columns={0:'category'})
dfcateg.head()

category_id              category
0              1              Hobbies 
71             2                   NaN
106            3  Characters & Brands 
282            4          Fancy Dress 
290            5                 Bags

In [11]:
dfsubc = pd.DataFrame(dfcate, columns=[1])
dfsubc.drop_duplicates(subset=None, keep='first', inplace=True)
dfsubc.drop(labels=None, axis=0, index=71, columns=None, level=None, inplace=False, errors='raise')
dfsubc.insert(0, 'subcategory_id', range(1, 1 + len(dfsubc)))
dfsubc = dfsubc.rename(columns={1:'subcategory'})
dfsubc.head()

subcategory_id                    subcategory
0                 1   Model Trains & Railway Sets 
71                2                            NaN
106               3              Thomas & Friends 
217               4                         Hasbro
236               5                     Star Wars

In [12]:
dfsubc1 = pd.DataFrame(dfcate, columns=[2])
dfsubc1.drop_duplicates(subset=None, keep='first', inplace=True)
dfsubc1.drop(labels=None, axis=0, index=71, columns=None, level=None, inplace=False, errors='raise')
dfsubc1.insert(0, 'subcategory1_id', range(1, 1 + len(dfsubc1)))
dfsubc1 = dfsubc1.rename(columns={2:'subcategory1'})
dfsubc1.head()

subcategory1_id                     subcategory1
0                  1                   Rail Vehicles 
5                  2   Lighting & Signal Engineering 
32                 3                     Accessories 
71                 4                              NaN
106                5                      Train Play

In [13]:
dfsubc2 = pd.DataFrame(dfcate, columns=[3])
dfsubc2.drop_duplicates(subset=None, keep='first', inplace=True)
dfsubc2.drop(labels=None, axis=0, index=71, columns=None, level=None, inplace=False, errors='raise')
dfsubc2.insert(0, 'subcategory2_id', range(1, 1 + len(dfsubc2)))
dfsubc2 = dfsubc2.rename(columns={3:'subcategory2'})
dfsubc2.head()

subcategory2_id              subcategory2
0                 1                    Trains
5                 2          Lamps & Lighting
10                3               Locomotives
29                4            Signal & Sound
32                5   Recreational Activities

In [14]:
dfdes = pd.DataFrame(df, columns=['product_description'])
dfdes.drop_duplicates(subset=None, keep='first', inplace=True)
dfdes.insert(0, 'description_id', range(1, 1 + len(dfdes)))
dfdes = dfdes.rename(columns={1:'product_description'})
dfdes.head()

description_id                                product_description
0               1  Product Description Hornby 2014 Catalogue Box ...
1               2  Size Name:Large FunkyBuys® Large Christmas Hol...
2               3  BIG CLASSIC TOY TRAIN SET TRACK CARRIAGE LIGHT...
3               4  Hornby 00 Gauge BR Hawksworth 3rd Class W 2107...
4               5  Product Description Hornby RailRoad 0-4-0 Gild...

In [15]:
dfinf = pd.DataFrame(df, columns=['product_information'])
dfinf.drop_duplicates(subset=None, keep='first', inplace=True)
dfinf.insert(0, 'information_id', range(1, 1 + len(dfinf)))
dfinf = dfinf.rename(columns={1:'product_information'})
dfinf.head()

information_id                                product_information
0               1  Technical Details Item Weight640 g Product Dim...
1               2  Technical Details Manufacturer recommended age...
2               3  Technical Details Manufacturer recommended age...
3               4  Technical Details Item Weight259 g Product Dim...
4               5  Technical Details Item Weight159 g Product Dim...

In [16]:
df['amazon_category_and_sub_category'] = df.amazon_category_and_sub_category.str.split(pat='>',expand=False)
dfcate = dfcate.rename(columns={0:'category',1:'subcategory',2:'subcategory1',3:'subcategory2'})
dfcate=dfcate.drop([4],axis=1)
dfcate.head()

category                    subcategory     subcategory1 subcategory2
0  Hobbies    Model Trains & Railway Sets    Rail Vehicles        Trains
1  Hobbies    Model Trains & Railway Sets    Rail Vehicles        Trains
2  Hobbies    Model Trains & Railway Sets    Rail Vehicles        Trains
3  Hobbies    Model Trains & Railway Sets    Rail Vehicles        Trains
4  Hobbies    Model Trains & Railway Sets    Rail Vehicles        Trains

In [17]:
df.drop(labels=None, axis=0, index=None, columns='amazon_category_and_sub_category', level=None, inplace=False, errors='raise')
df.join(dfcate)

product_id                                       product_name  \
0              1                              Hornby 2014 Catalogue   
1              2  FunkyBuys® Large Christmas Holiday Express Fes...   
2              3  CLASSIC TOY TRAIN SET TRACK CARRIAGES LIGHT EN...   
3              4     HORNBY Coach R4410A BR Hawksworth Corridor 3rd   
4              5  Hornby 00 Gauge 0-4-0 Gildenlow Salt Co. Steam...   
...          ...                                                ...   
9995        9993  Batman 1966 TV Series Action Figures - The Rid...   
9996        9994  Star Wars Costume, Kids Stormtrooper Costume S...   
9997        9995        Defiance Lawkeeper Metal Badge Prop Replica   
9998        9996  Justice League of America Series 3 Green Lante...   
9999        9997                  Star Wars 1/72 Y-Wing Starfighter   

                 manufacturer  price number_available_in_stock  \
0                      Hornby   3.42                     5 new   
1                   FunkyBuys  16.99                       NaN   
2                         ccf   9.99                     2 new   
3                      Hornby  39.99                       NaN   
4                      Hornby  32.19                       NaN   
...                       ...    ...                       ...   
9995                   Mattel  22.95                     5 new   
9996                Star Wars  39.99                       NaN   
9997  Olde Scotland Yard Ltd.  43.99                     3 new   
9998                DC Comics  49.81                     3 new   
9999                   Bandai  21.20                    31 new   

     number_of_reviews average_review_rating  \
0                   15    4.9 out of 5 stars   
1                    2    4.5 out of 5 stars   
2                   17    3.9 out of 5 stars   
3                    1    5.0 out of 5 stars   
4                    3    4.7 out of 5 stars   
...                ...                   ...   
9995                 3    5.0 out of 5 stars   
9996                 1    4.0 out of 5 stars   
9997                 1    5.0 out of 5 stars   
9998                 1    5.0 out of 5 stars   
9999                11    4.5 out of 5 stars   

                       amazon_category_and_sub_category  \
0     [Hobbies ,  Model Trains & Railway Sets ,  Rai...   
1     [Hobbies ,  Model Trains & Railway Sets ,  Rai...   
2     [Hobbies ,  Model Trains & Railway Sets ,  Rai...   
3     [Hobbies ,  Model Trains & Railway Sets ,  Rai...   
4     [Hobbies ,  Model Trains & Railway Sets ,  Rai...   
...                                                 ...   
9995  [Hobbies ,  Collectible Figures & Memorabilia ...   
9996         [Characters & Brands ,  Star Wars ,  Toys]   
9997  [Novelty & Special Use ,  Novelty ,  Accessori...   
9998  [Hobbies ,  Collectible Figures & Memorabilia ...   
9999         [Characters & Brands ,  Star Wars ,  Toys]   

             customers_who_bought_this_item_also_bought  \
0     http://www.amazon.co.uk/Hornby-R8150-Catalogue...   
1     http://www.amazon.co.uk/Christmas-Holiday-Expr...   
2     http://www.amazon.co.uk/Classic-Train-Lights-B...   
3                                                   NaN   
4     http://www.amazon.co.uk/Hornby-R6367-RailRoad-...   
...                                                 ...   
9995  http://www.amazon.co.uk/Batman-1966-Series-Act...   
9996  http://www.amazon.co.uk/Storm-Trooper-Costume-...   
9997                                                NaN   
9998                                                NaN   
9999  http://www.amazon.co.uk/Bandai-Star-Wars-Fight...   

                                    product_information  \
0     Technical Details Item Weight640 g Product Dim...   
1     Technical Details Manufacturer recommended age...   
2     Technical Details Manufacturer recommended age...   
3     Technical Details Item Weight259 g Product Dim...   
4     Technical Details Item Weight159 g Product Dim...   
...        

In [18]:
dfstock = pd.DataFrame(df, columns=['number_available_in_stock'])
dfstock = dfstock.fillna(0)
dfstock = dfstock.number_available_in_stock.str.split(n=1,expand=True)
dfstock.head()

0    1
0    5  new
1  NaN  NaN
2    2  new
3  NaN  NaN
4  NaN  NaN

In [19]:
dfstocknum = pd.DataFrame(dfstock, columns=[0])
dfstocknum = dfstocknum.fillna(0)
dfstocknum =dfstocknum.rename(columns={0:'stocknumber'})
dfstocknum.head()

stocknumber
0           5
1           0
2           2
3           0
4           0

In [20]:
dfstocksta = pd.DataFrame(dfstock, columns=[1])
dfstocksta = dfstocksta.fillna('None')
dfstocksta = dfstocksta.rename(columns={1:'stockstatus'})
dfstock = dfstocknum.join(dfstocksta)
df = df.join(dfstock)
df.head()

product_id                                       product_name manufacturer  \
0           1                              Hornby 2014 Catalogue       Hornby   
1           2  FunkyBuys® Large Christmas Holiday Express Fes...    FunkyBuys   
2           3  CLASSIC TOY TRAIN SET TRACK CARRIAGES LIGHT EN...          ccf   
3           4     HORNBY Coach R4410A BR Hawksworth Corridor 3rd       Hornby   
4           5  Hornby 00 Gauge 0-4-0 Gildenlow Salt Co. Steam...       Hornby   

   price number_available_in_stock number_of_reviews average_review_rating  \
0   3.42                     5 new                15    4.9 out of 5 stars   
1  16.99                       NaN                 2    4.5 out of 5 stars   
2   9.99                     2 new                17    3.9 out of 5 stars   
3  39.99                       NaN                 1    5.0 out of 5 stars   
4  32.19                       NaN                 3    4.7 out of 5 stars   

                    amazon_category_and_sub_category  \
0  [Hobbies ,  Model Trains & Railway Sets ,  Rai...   
1  [Hobbies ,  Model Trains & Railway Sets ,  Rai...   
2  [Hobbies ,  Model Trains & Railway Sets ,  Rai...   
3  [Hobbies ,  Model Trains & Railway Sets ,  Rai...   
4  [Hobbies ,  Model Trains & Railway Sets ,  Rai...   

          customers_who_bought_this_item_also_bought  \
0  http://www.amazon.co.uk/Hornby-R8150-Catalogue...   
1  http://www.amazon.co.uk/Christmas-Holiday-Expr...   
2  http://www.amazon.co.uk/Classic-Train-Lights-B...   
3                                                NaN   
4  http://www.amazon.co.uk/Hornby-R6367-RailRoad-...   

                                 product_information  \
0  Technical Details Item Weight640 g Product Dim...   
1  Technical Details Manufacturer recommended age...   
2  Technical Details Manufacturer recommended age...   
3  Technical Details Item Weight259 g Product Dim...   
4  Technical Details Item Weight159 g Product Dim...   

                                 product_description  \
0  Product Description Hornby 2014 Catalogue Box ...   
1  Size Name:Large FunkyBuys® Large Christmas Hol...   
2  BIG CLASSIC TOY TRAIN SET TRACK CARRIAGE LIGHT...   
3  Hornby 00 Gauge BR Hawksworth 3rd Class W 2107...   
4  Product Description Hornby RailRoad 0-4-0 Gild...   

         items_customers_buy_after_viewing_this_item  \
0  http://www.amazon.co.uk/Hornby-R8150-Catalogue...   
1  http://www.amazon.co.uk/Christmas-Holiday-Expr...   
2  http://www.amazon.co.uk/Train-With-Tracks-Batt...   
3                                                NaN   
4  http://www.amazon.co.uk/Hornby-R2672-RailRoad-...   

                      customer_questions_and_answers  \
0  Does this catalogue detail all the previous Ho...   
1  can you turn off sounds // hi no you cant turn...   
2  What is the gauge of the track // Hi Paul.Trut...   
3                                                NaN   
4                                                NaN   

                                    customer_reviews stocknumber stockstatus  
0  Worth Buying For The Pictures Alone (As Ever) ...           5         new  
1  Four Stars // 4.0 // 18 Dec. 2015 // By\n    \...           0        None  
2  **Highly Recommended!** // 5.0 // 26 May 2015 ...           2         new  
3  I love it // 5.0 // 22 July 2013 // By\n    \n...           0        None  
4  Birthday present // 5.0 // 14 April 2014 // By...           0        None

In [21]:
df=df.drop(['number_available_in_stock','amazon_category_and_sub_category'],axis=1)
df = df.join(dfcate)
df.head()

product_id                                       product_name manufacturer  \
0           1                              Hornby 2014 Catalogue       Hornby   
1           2  FunkyBuys® Large Christmas Holiday Express Fes...    FunkyBuys   
2           3  CLASSIC TOY TRAIN SET TRACK CARRIAGES LIGHT EN...          ccf   
3           4     HORNBY Coach R4410A BR Hawksworth Corridor 3rd       Hornby   
4           5  Hornby 00 Gauge 0-4-0 Gildenlow Salt Co. Steam...       Hornby   

   price number_of_reviews average_review_rating  \
0   3.42                15    4.9 out of 5 stars   
1  16.99                 2    4.5 out of 5 stars   
2   9.99                17    3.9 out of 5 stars   
3  39.99                 1    5.0 out of 5 stars   
4  32.19                 3    4.7 out of 5 stars   

          customers_who_bought_this_item_also_bought  \
0  http://www.amazon.co.uk/Hornby-R8150-Catalogue...   
1  http://www.amazon.co.uk/Christmas-Holiday-Expr...   
2  http://www.amazon.co.uk/Classic-Train-Lights-B...   
3                                                NaN   
4  http://www.amazon.co.uk/Hornby-R6367-RailRoad-...   

                                 product_information  \
0  Technical Details Item Weight640 g Product Dim...   
1  Technical Details Manufacturer recommended age...   
2  Technical Details Manufacturer recommended age...   
3  Technical Details Item Weight259 g Product Dim...   
4  Technical Details Item Weight159 g Product Dim...   

                                 product_description  \
0  Product Description Hornby 2014 Catalogue Box ...   
1  Size Name:Large FunkyBuys® Large Christmas Hol...   
2  BIG CLASSIC TOY TRAIN SET TRACK CARRIAGE LIGHT...   
3  Hornby 00 Gauge BR Hawksworth 3rd Class W 2107...   
4  Product Description Hornby RailRoad 0-4-0 Gild...   

         items_customers_buy_after_viewing_this_item  \
0  http://www.amazon.co.uk/Hornby-R8150-Catalogue...   
1  http://www.amazon.co.uk/Christmas-Holiday-Expr...   
2  http://www.amazon.co.uk/Train-With-Tracks-Batt...   
3                                                NaN   
4  http://www.amazon.co.uk/Hornby-R2672-RailRoad-...   

                      customer_questions_and_answers  \
0  Does this catalogue detail all the previous Ho...   
1  can you turn off sounds // hi no you cant turn...   
2  What is the gauge of the track // Hi Paul.Trut...   
3                                                NaN   
4                                                NaN   

                                    customer_reviews stocknumber stockstatus  \
0  Worth Buying For The Pictures Alone (As Ever) ...           5         new   
1  Four Stars // 4.0 // 18 Dec. 2015 // By\n    \...           0        None   
2  **Highly Recommended!** // 5.0 // 26 May 2015 ...           2         new   
3  I love it // 5.0 // 22 July 2013 // By\n    \n...           0        None   
4  Birthday present // 5.0 // 14 April 2014 // By...           0        None   

   category                    subcategory     subcategory1 subcategory2  
0  Hobbies    Model Trains & Railway Sets    Rail Vehicles        Trains  
1  Hobbies    Model Trains & Railway Sets    Rail Vehicles        Trains  
2  Hobbies    Model Trains & Railway Sets    Rail Vehicles        Trains  
3  Hobbies    Model Trains & Railway Sets    Rail Vehicles        Trains  
4  Hobbies    Model Trains & Railway Sets    Rail Vehicles        Trains

In [22]:
dfstocksta.drop_duplicates(subset=None, keep='first', inplace=True)
dfstocksta.insert(0,'stockstatus_id', range(1, 1 + len(dfstocksta)))
dfstocksta.head()

stockstatus_id  stockstatus
0                  1          new
1                  2         None
17                 3         used
812                4  collectible
1079               5  refurbished

In [23]:
#dfas = df.customers_who_bought_this_item_also_bought.str.split('|').apply(pd.Series, 1).stack()
#dfas.index = dfas.index.droplevel(-1)
#dfas.name = 'customers_who_bought_this_item_also_bought'
#del df['customers_who_bought_this_item_also_bought']
#df = df.join(dfas)

In [24]:
#dfview = df.items_customers_buy_after_viewing_this_item.str.split('|').apply(pd.Series, 1).stack()
#dfview.index = dfview.index.droplevel(-1)
#dfview.name = 'items_customers_buy_after_viewing_this_item'
#del df['items_customers_buy_after_viewing_this_item']
#df = df.join(dfview)

In [25]:
df = df.reset_index()

In [26]:
df.drop(labels=None, axis=0, index=None, columns='index', level=None, inplace=False, errors='raise')

product_id                                       product_name  \
0              1                              Hornby 2014 Catalogue   
1              2  FunkyBuys® Large Christmas Holiday Express Fes...   
2              3  CLASSIC TOY TRAIN SET TRACK CARRIAGES LIGHT EN...   
3              4     HORNBY Coach R4410A BR Hawksworth Corridor 3rd   
4              5  Hornby 00 Gauge 0-4-0 Gildenlow Salt Co. Steam...   
...          ...                                                ...   
9992        9993  Batman 1966 TV Series Action Figures - The Rid...   
9993        9994  Star Wars Costume, Kids Stormtrooper Costume S...   
9994        9995        Defiance Lawkeeper Metal Badge Prop Replica   
9995        9996  Justice League of America Series 3 Green Lante...   
9996        9997                  Star Wars 1/72 Y-Wing Starfighter   

                 manufacturer  price number_of_reviews average_review_rating  \
0                      Hornby   3.42                15    4.9 out of 5 stars   
1                   FunkyBuys  16.99                 2    4.5 out of 5 stars   
2                         ccf   9.99                17    3.9 out of 5 stars   
3                      Hornby  39.99                 1    5.0 out of 5 stars   
4                      Hornby  32.19                 3    4.7 out of 5 stars   
...                       ...    ...               ...                   ...   
9992                   Mattel  22.95                 3    5.0 out of 5 stars   
9993                Star Wars  39.99                 1    4.0 out of 5 stars   
9994  Olde Scotland Yard Ltd.  43.99                 1    5.0 out of 5 stars   
9995                DC Comics  49.81                 1    5.0 out of 5 stars   
9996                   Bandai  21.20                11    4.5 out of 5 stars   

             customers_who_bought_this_item_also_bought  \
0     http://www.amazon.co.uk/Hornby-R8150-Catalogue...   
1     http://www.amazon.co.uk/Christmas-Holiday-Expr...   
2     http://www.amazon.co.uk/Classic-Train-Lights-B...   
3                                                   NaN   
4     http://www.amazon.co.uk/Hornby-R6367-RailRoad-...   
...                                                 ...   
9992  http://www.amazon.co.uk/Batman-1966-Series-Act...   
9993  http://www.amazon.co.uk/Storm-Trooper-Costume-...   
9994                                                NaN   
9995                                                NaN   
9996  http://www.amazon.co.uk/Bandai-Star-Wars-Fight...   

                                    product_information  \
0     Technical Details Item Weight640 g Product Dim...   
1     Technical Details Manufacturer recommended age...   
2     Technical Details Manufacturer recommended age...   
3     Technical Details Item Weight259 g Product Dim...   
4     Technical Details Item Weight159 g Product Dim...   
...                                                 ...   
9992  Technical Details Item Weight136 g Product Dim...   
9993  Technical Details    Additional Information AS...   
9994  Technical Details Item Weight159 g Product Dim...   
9995  Technical Details Item Weight181 g Product Dim...   
9996  Technical Details Item Weight299 g Product Dim...   

                                    product_description  \
0     Product Description Hornby 2014 Catalogue Box ...   
1     Size Name:Large FunkyBuys® Large Christmas Hol...   
2     BIG CLASSIC TOY TRAIN SET TRACK CARRIAGE LIGHT...   
3     Hornby 00 Gauge BR Hawksworth 3rd Class W 2107...   
4     Product Description Hornby RailRoad 0-4-0 Gild...   
...                                                 ...   
9992  DC 66 Batman Classic TV Series 6 Inch Riddler ...   
9993                                                NaN   
9994  *Includes 1x Badge with holder and chain *High...   
9995  Designed by Ed BenesIt's here - the third seri...   
9996                                                NaN   

            items_customers_buy_after_viewing_this_item  \
0     

In [28]:
dfmaf = pd.DataFrame(df, columns=['manufacturer'])
dfmaf.drop_duplicates(subset=None, keep='first', inplace=True)
dfmaf.insert(0, 'manufacturer_id', range(1, 1 + len(dfmaf)))
dfmaf.head()

manufacturer_id manufacturer
0                1       Hornby
1                2    FunkyBuys
2                3          ccf
5                4      Generic
9                5  Chuggington

In [29]:
temp1 = df
temp2 = dfmaf
temp1 = temp1.merge(temp2, on='manufacturer', how='left', indicator=False)

In [30]:
temp1 = temp1.merge(dfcateg, on='category', how='left', indicator=False)
temp1 = temp1.merge(dfdes, on='product_description', how='left', indicator=False)
temp1 = temp1.merge(dfinf, on='product_information', how='left', indicator=False)
temp1 = temp1.merge(dfstocksta, on='stockstatus', how='left', indicator=False)
temp1=temp1.drop(['manufacturer','product_information','product_description','category','stockstatus','subcategory','subcategory1','subcategory2','index'],axis=1)

In [31]:
temp1["number_of_reviews"] = pd.to_numeric(temp1["number_of_reviews"],errors='coerce')

In [32]:
df['stocknumber'] = df['stocknumber'].str.extractall('(\d+)').unstack().fillna('').sum(axis=1).astype(int)

In [33]:
temp1

product_id                                       product_name  price  \
0              1                              Hornby 2014 Catalogue   3.42   
1              2  FunkyBuys® Large Christmas Holiday Express Fes...  16.99   
2              3  CLASSIC TOY TRAIN SET TRACK CARRIAGES LIGHT EN...   9.99   
3              4     HORNBY Coach R4410A BR Hawksworth Corridor 3rd  39.99   
4              5  Hornby 00 Gauge 0-4-0 Gildenlow Salt Co. Steam...  32.19   
...          ...                                                ...    ...   
9992        9993  Batman 1966 TV Series Action Figures - The Rid...  22.95   
9993        9994  Star Wars Costume, Kids Stormtrooper Costume S...  39.99   
9994        9995        Defiance Lawkeeper Metal Badge Prop Replica  43.99   
9995        9996  Justice League of America Series 3 Green Lante...  49.81   
9996        9997                  Star Wars 1/72 Y-Wing Starfighter  21.20   

      number_of_reviews average_review_rating  \
0                  15.0    4.9 out of 5 stars   
1                   2.0    4.5 out of 5 stars   
2                  17.0    3.9 out of 5 stars   
3                   1.0    5.0 out of 5 stars   
4                   3.0    4.7 out of 5 stars   
...                 ...                   ...   
9992                3.0    5.0 out of 5 stars   
9993                1.0    4.0 out of 5 stars   
9994                1.0    5.0 out of 5 stars   
9995                1.0    5.0 out of 5 stars   
9996               11.0    4.5 out of 5 stars   

             customers_who_bought_this_item_also_bought  \
0     http://www.amazon.co.uk/Hornby-R8150-Catalogue...   
1     http://www.amazon.co.uk/Christmas-Holiday-Expr...   
2     http://www.amazon.co.uk/Classic-Train-Lights-B...   
3                                                   NaN   
4     http://www.amazon.co.uk/Hornby-R6367-RailRoad-...   
...                                                 ...   
9992  http://www.amazon.co.uk/Batman-1966-Series-Act...   
9993  http://www.amazon.co.uk/Storm-Trooper-Costume-...   
9994                                                NaN   
9995                                                NaN   
9996  http://www.amazon.co.uk/Bandai-Star-Wars-Fight...   

            items_customers_buy_after_viewing_this_item  \
0     http://www.amazon.co.uk/Hornby-R8150-Catalogue...   
1     http://www.amazon.co.uk/Christmas-Holiday-Expr...   
2     http://www.amazon.co.uk/Train-With-Tracks-Batt...   
3                                                   NaN   
4     http://www.amazon.co.uk/Hornby-R2672-RailRoad-...   
...                                                 ...   
9992                                                NaN   
9993  http://www.amazon.co.uk/Storm-Trooper-Costume-...   
9994  http://www.amazon.co.uk/Olde-Scotland-Yard-Ltd...   
9995                                                NaN   
9996                                                NaN   

                         customer_questions_and_answers  \
0     Does this catalogue detail all the previous Ho...   
1     can you turn off sounds // hi no you cant turn...   
2     What is the gauge of the track // Hi Paul.Trut...   
3                                                   NaN   
4                                                   NaN   
...                                                 ...   
9992                                                NaN   
9993                                                NaN   
9994                                                NaN   
9995                                                NaN   
9996                                                NaN   

                                       customer_reviews stocknumber  \
0     Worth Buying For The Pictures Alone (As Ever) ...           5   
1     Four Stars // 4.0 // 18 Dec. 2015 // By\n    \...           0   
2     **Highly Recommended!** // 5.0 // 26 May 2015 ...           2   
3     I love it // 5.0 // 22 July 2013 // By\n    \n...           0  

In [34]:
dfcate.insert(0, 'category_id(PK)', range(1, 1 + len(dfcate)))
dfcate

category_id(PK)                category  \
0                   1                Hobbies    
1                   2                Hobbies    
2                   3                Hobbies    
3                   4                Hobbies    
4                   5                Hobbies    
...               ...                     ...   
9995             9993                Hobbies    
9996             9994    Characters & Brands    
9997             9995  Novelty & Special Use    
9998             9996                Hobbies    
9999             9997    Characters & Brands    

                              subcategory                      subcategory1  \
0            Model Trains & Railway Sets                     Rail Vehicles    
1            Model Trains & Railway Sets                     Rail Vehicles    
2            Model Trains & Railway Sets                     Rail Vehicles    
3            Model Trains & Railway Sets                     Rail Vehicles    
4            Model Trains & Railway Sets                     Rail Vehicles    
...                                   ...                               ...   
9995   Collectible Figures & Memorabilia    Collectible Props & Memorabilia   
9996                           Star Wars                               Toys   
9997                             Novelty                       Accessories    
9998   Collectible Figures & Memorabilia    Collectible Props & Memorabilia   
9999                           Star Wars                               Toys   

         subcategory2  
0              Trains  
1              Trains  
2              Trains  
3              Trains  
4              Trains  
...               ...  
9995             None  
9996             None  
9997   Buttons & Pins  
9998             None  
9999             None  

[9997 rows x 5 columns]

In [35]:
temp2 = dfcate
temp2 = temp2.merge(dfsubc, on='subcategory', how='left', indicator=False)
temp2 = temp2.merge(dfcateg, on='category', how='left', indicator=False)
temp2 = temp2.drop(['category','subcategory','subcategory1','subcategory2','category_id(PK)'],axis =1)


In [36]:

temp2 = temp2.rename({'category_id(PK)_y':'category_id(FK)','subcategory_id(PK)':'subcategory_id(FK)'},axis=1)
temp2 = temp2.drop_duplicates()
temp2

subcategory_id  category_id
0                  1            1
71                 2            2
106                3            3
217                4            3
236                5            3
...              ...          ...
9446             162           22
9468             163           21
9726             164           44
9909             165            8
9929             166            8

[170 rows x 2 columns]

In [37]:
temp3 = dfcate
temp3 = temp3.drop(['category'],axis =1)
temp3 = temp3.merge(dfsubc1, on='subcategory1', how='left', indicator=False)
temp3 = temp3.merge(dfsubc, on='subcategory', how='left', indicator=False)
temp3 = temp3.drop(['subcategory','subcategory1','subcategory2','category_id(PK)'],axis =1)
temp3 = temp3.drop_duplicates()
temp2 = temp2.rename({'subcategory_id(PK)':'subcategory_id(FK)','subcategory1_id(PK)':'subcategory1_id(FK)'},axis=1)
temp3

subcategory1_id  subcategory_id
0                   1               1
5                   2               1
32                  3               1
71                  4               2
106                 5               3
...               ...             ...
9567              114             149
9637              115              33
9726              116             164
9909                4             165
9929                4             166

[229 rows x 2 columns]

In [38]:
temp4 = dfcate
temp4 = temp4.merge(dfsubc2, on='subcategory2', how='left', indicator=False)
temp4 = temp4.merge(dfsubc1, on='subcategory1', how='left', indicator=False)
temp4 = temp4.drop(['category','subcategory','subcategory1','subcategory2','category_id(PK)'],axis =1)
temp4 = temp4.drop_duplicates()
temp4 = temp4.rename({'subcategory1_id(PK)':'subcategory1_id(FK)','subcategory2_id(PK)':'subcategory2_id(FK)'},axis=1)
temp4

subcategory2_id  subcategory1_id
0                   1                1
5                   2                2
10                  3                1
29                  4                2
32                  5                3
...               ...              ...
9492                7              113
9567                7              114
9610               47               70
9637                7              115
9726                7              116

[139 rows x 2 columns]

## Create database tables

Based on the normalization plan we are now ready to create the database tables. We will firstly create a product database in postgresql.

In [41]:
conn_url = 'postgresql://postgres:123@localhost/555'

In [42]:
engine = create_engine(conn_url)

In [43]:
connection = engine.connect()

In [44]:
stmt = """
    CREATE TABLE manufacturer (
        manufacturer_id      integer,
        manufacturer   varchar(100),
        PRIMARY KEY (manufacturer_id)
    
    );
    
    CREATE TABLE category (
        category_id      integer,
        category   varchar(100) ,
        PRIMARY KEY (category_id)
    
    );
    
    CREATE TABLE subcategory (
        subcategory_id      integer,
        subcategory   varchar(100),
        PRIMARY KEY (subcategory_id)
    
    );
    
    CREATE TABLE subcategory1 (
        subcategory1_id      integer,
        subcategory1   varchar(100),
        PRIMARY KEY (subcategory1_id)
    
    );
    
    CREATE TABLE subcategory2 (
        subcategory2_id      integer,
        subcategory2   varchar(100),
        PRIMARY KEY (subcategory2_id)
    
    );
    
    CREATE TABLE product_description (
        description_id      integer,
        product_description   varchar(8000),
        PRIMARY KEY (description_id)
    
    );
    
     CREATE TABLE product_information (
        information_id      integer,
        product_information   varchar(8000),
        PRIMARY KEY (information_id)
    
    );
    
    CREATE TABLE stockstatus (
        stockstatus_id      integer,
        stockstatus   varchar(100),
        PRIMARY KEY (stockstatus_id)
    
    );
    
    
    CREATE TABLE category_subcategory (
        category_id      integer,
        subcategory_id      integer,
        PRIMARY KEY (category_id,subcategory_id)
    );
    
     
    CREATE TABLE subcategory_subcategory1 (
        subcategory_id     integer,
        subcategory1_id      integer,
        PRIMARY KEY (subcategory_id,subcategory1_id)
    );
    
     
    CREATE TABLE subcategory1_subcategory2 (
        subcategory1_id      integer,
        subcategory2_id      integer,
        PRIMARY KEY (subcategory1_id,subcategory2_id)
    );
    
   
"""
connection.execute(stmt)

In [45]:
stmt = """
    CREATE TABLE product (
            product_id   integer,
            product_name    varchar(5000) NOT NULL,
            price     float,
            number_of_reviews         integer,
            average_review_rating    varchar(50),
            customers_who_bought_this_item_also_bought    varchar(8000),
            items_customers_buy_after_viewing_this_item     varchar(8000),
            customer_questions_and_answers    varchar(8000),
            customer_reviews    varchar(8000),
            stocknumber   integer NOT NULL,
            manufacturer_id    integer,
            category_id    integer,
            description_id    integer,
            information_id    integer,
            stockstatus_id    integer,
            PRIMARY KEY (product_id,price),
            FOREIGN KEY (manufacturer_id) REFERENCES manufacturer (manufacturer_id),
            FOREIGN KEY (category_id) REFERENCES category (category_id),
            FOREIGN KEY (description_id) REFERENCES product_description (description_id),
            FOREIGN KEY (information_id) REFERENCES product_information (information_id)
        
    );

"""  
connection.execute(stmt)

In [46]:
#Create table for the future use
stmt = """ 
    create table company(
        name           varchar(255) not null
            primary key,
        department_sum int          not null,
        employees_sum  int          not null,
        constraint company_name_uindex
            unique (name)
    );

    create table employees(
            id            int 
                primary key,
            name          varchar(80) not null,
            gender        varchar(10) not null,
            phone_number  varchar(80) not null,
            department_id int         not null,
            salary        double precision      not null,
            result_id     int         null,
            constraint employees_id_uindex
                unique (id),
            constraint employees_department_id_fk
                foreign key (department_id) references department (id)
    );
    create table up_department(
        id          int    not null
            primary key,
        items_customers_buy_after_viewing_this_item               varchar(500) not null,
        down_department_id int          null
    );
    create table department(
        id               int 
            primary key,
        employees_sum    int         not null,
        category         varchar(80) null,
        leader_id        int         null,
        up_department_id int         not null,
        constraint department_id_uindex
            unique (id),
        constraint department_leader_id_fk
            foreign key (leader_id) references leader (id),
        constraint department_up_department_id_fk
            foreign key (up_department_id) references up_department (id)
    );



"""

connection.execute(stmt)

ProgrammingError: (psycopg2.errors.UndefinedTable) relation "department" does not exist

[SQL:  
    create table company(
        name           varchar(255) not null
            primary key,
        department_sum int          not null,
        employees_sum  int          not null,
        constraint company_name_uindex
            unique (name)
    );

    create table employees(
            id            int 
                primary key,
            name          varchar(80) not null,
            gender        varchar(10) not null,
            phone_number  varchar(80) not null,
            department_id int         not null,
            salary        double precision      not null,
            result_id     int         null,
            constraint employees_id_uindex
                unique (id),
            constraint employees_department_id_fk
                foreign key (department_id) references department (id)
    );
    create table up_department(
        id          int    not null
            primary key,
        items_customers_buy_after_viewing_this_item               varchar(500) not null,
        down_department_id int          null
    );
    create table department(
        id               int 
            primary key,
        employees_sum    int         not null,
        category         varchar(80) null,
        leader_id        int         null,
        up_department_id int         not null,
        constraint department_id_uindex
            unique (id),
        constraint department_leader_id_fk
            foreign key (leader_id) references leader (id),
        constraint department_up_department_id_fk
            foreign key (up_department_id) references up_department (id)
    );



]
(Background on this error at: https://sqlalche.me/e/14/f405)

## ETL Process

Because in the step Normalization we have already prepared well about the data and the table, so we just need a few more steps to finish extracting and transforming the data and then directly load the data into our database.

In [47]:
dfmaf.to_sql(name='manufacturer', con=engine, if_exists='append', index=False)

652

In [48]:
dfcateg.to_sql(name='category', con=engine, if_exists='append', index=False)

44

In [49]:
dfsubc.to_sql(name='subcategory', con=engine, if_exists='append', index=False)

166

In [50]:
dfsubc1.to_sql(name='subcategory1', con=engine, if_exists='append', index=False)

116

In [51]:
dfsubc2.to_sql(name='subcategory2', con=engine, if_exists='append', index=False)

47

In [52]:
dfinf.to_sql(name='product_information', con=engine, if_exists='append', index=False)

940

In [53]:
dfstocksta.to_sql(name='stockstatus', con=engine, if_exists='append', index=False)

5

In [54]:
#The length of the value exceed the maximum constraint(Varchar(5000))
dfdes.to_sql(name='product_description', con=engine, if_exists='append', index=False)
temp1.to_sql(name='product', con=engine, if_exists='append', index=False)

DataError: (psycopg2.errors.StringDataRightTruncation) value too long for type character varying(8000)

[SQL: INSERT INTO product_description (description_id, product_description) VALUES (%(description_id)s, %(product_description)s)]
[parameters: ({'description_id': 1, 'product_description': 'Product Description Hornby 2014 Catalogue Box Contains 1 x one catalogue'}, {'description_id': 2, 'product_description': 'Size Name:Large FunkyBuys® Large Christmas Holiday Express Festive Train Set (SI-TY1017) Toy Light / Sounds / Battery Operated & Smoke'}, {'description_id': 3, 'product_description': 'BIG CLASSIC TOY TRAIN SET TRACK CARRIAGE LIGHT ENGINE SOUND BOXED KIDS BATTERY Railway Train Set with Light Sound, Big Size Curved Track Free Wheelin ... (59 characters truncated) ...  Assembled In More Layouts Length Of Track 104 cm WIDTH OF THE TRACK 68 CM Finely Detailed Realistic Toy Train REQUIRES 2 AA BATTERIES (NOT INCLUDED)'}, {'description_id': 4, 'product_description': 'Hornby 00 Gauge BR Hawksworth 3rd Class W 2107 W # R4410A'}, {'description_id': 5, 'product_description': 'Product Description Hornby RailRoad 0-4-0 Gildenlow Salt Co 00 gauge steam locomotive model. Safety warning: This product is not suitable for childre ... (100 characters truncated) ... nctional sharp edges. Handle with care. Only use this product with the recommended transformer. Made in China. Box Contains 1x Steam Locomotive Model'}, {'description_id': 6, 'product_description': 'These delicate model garden lights are mainly used in teaching, photography, and various kinds of scene model. Each of them is completed with wires a ... (482 characters truncated) ... e power, please connect two long wires with one end and connect the rest two short wires with another end. Package Include: 20pcs model garden lights'}, {'description_id': 7, 'product_description': 'Product Description Hornby BR bogie passenger brake coach has pristine finish. Livery: BR red livery. Entered Service: 1930. Period: 1950. 230mm coac ... (74 characters truncated) ... ffers something for most enthusiasts modelling the various regions and periods of the British rail network. Box Contains 1x Red Bogie Passenger Brake'}, {'description_id': 8, 'product_description': "Product Description Inject a bit of Hornby magic into Christmas with the special Santa's Express Train Set. The set includes everything you need to g ... (1321 characters truncated) ...  Box Van 1x 3rd Radius Starter Oval 1x Power Track and Track Straight 1xR8250 Train Controller 1x P9000 Transformer 1x Hornby MidiMat (1600 x 1180mm)"}  ... displaying 10 of 8515 total bound parameter sets ...  {'description_id': 8514, 'product_description': '*Includes 1x Badge with holder and chain *High quality metal construction *Ball clasp necklace *Removable from holster, features heavy duty pin design *Brand new'}, {'description_id': 8515, 'product_description': 'Designed by Ed BenesIt\'s here - the third series based on the popular DC Comics series, JUSTICE LEAGUE OF AMERICA, written by New York Times best-se ... (137 characters truncated) ... be made into action figures!- Green Lantern - 6.75" HEach figure features multiple points of articulation and a base. 4-color blister card packaging.'})]
(Background on this error at: https://sqlalche.me/e/14/9h9h)

In [ ]:
#For the future data
#employees.to_sql(name='employees', con=engine, if_exists='replace', index=False)
#department.to_sql(name='department', con=engine, if_exists='replace', index=False)
#up_department = df[['id','product_id','items_customers_buy_after_viewing_this_item']]
#up_department.to_sql(name='up_department_df', con=engine, if_exists='replace', index=False)

## Reasoning

We first looked at this dataset and found that it was in a file format that could be read directly in Python. So we performed a data inspection using the tool Jupyter Notebook. After a series of operations were performed, we would establish a connection to the SQL database in Python. Finally, we performed the ETL process to finalize the database.